### PR TITLE
test: add sources directory cleaning test

### PR DESCRIPTION
**Pull request by rjbou**
related #5448

### CODE REVIEW

File: tests/reftests/clean.test [hunk 1]
```diff
@@ -108,3 +109,160 @@ Remaining files:
...
+# Return code 1 #
+### opam remove waste
+The following actions will be performed:
+=== remove 1 package
+  - remove waste 1
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> removed   waste.1
+Done.
+### test -d OPAM/clean/.opam-switch/sources/waste
```
Comment (kit-ty-kate) [inline tests/reftests/clean.test:153]:
```suggestion
### ls OPAM/clean/.opam-switch/sources
```

File: tests/reftests/clean.test [hunk 2]
```diff
@@ -109,3 +110,175 @@ Remaining files:
...
+The following actions will be performed:
+=== install 1 package
+  - install waste 1
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> retrieved waste.1  (file://${BASEDIR}/waste.tgz)
+-> installed waste.1
+Done.
+### ls OPAM/clean/.opam-switch/sources/
+waste.1
```
Comment (dra27) [inline tests/reftests/clean.test:158]: This one's part of the original bug, right? After opam install, there's no need for it to be there?

File: tests/reftests/clean.test [hunk 3]
```diff
@@ -109,3 +110,175 @@ Remaining files:
...
+The following actions will be performed:
+=== recompile 1 package
+  - recompile waste 1
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> retrieved waste.1  (cached)
+-> removed   waste.1
+-> installed waste.1
+Done.
+### ls OPAM/clean/.opam-switch/sources
```
Comment (dra27) [inline tests/reftests/clean.test:187]: Correct the difference here is that `opam install waste` has left the sources directory populated, but `opam reinstall waste` has removed it?

File: tests/reftests/clean.test [hunk 4]
```diff
@@ -109,3 +110,175 @@ Remaining files:
...
+### opam remove waste
+The following actions will be performed:
+=== remove 1 package
+  - remove waste 1
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> removed   waste.1
+Done.
+### opam install ./waste
```
Comment (dra27) [inline tests/reftests/clean.test:196]:
```suggestion
### ls OPAM/clean/.opam-switch/sources
### opam install ./waste
```
to be sure that the `opam remove` didn't re-extract, or unneccessary? It seems worth terminating each test case with the explicit ls?

File: tests/reftests/clean.test [hunk 5]
```diff
@@ -109,3 +110,175 @@ Remaining files:
...
+=== install 1 package
+  - install waste 1 (pinned)
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> retrieved waste.1  (file://${BASEDIR}/waste)
+-> installed waste.1
+Done.
+### ls OPAM/clean/.opam-switch/sources
+waste
+### opam unpin waste
```
Comment (dra27) [inline tests/reftests/clean.test:241]: It feels like there should be more `ls` checks after these tests, confirming that exactly when the `waste` directory should still be there?

**Review by dra27 — COMMENTED** [review]
Apparently for me it's much easier to suggest to _other_ people to add comments to their tests, than it is to remember to put comments in _my_ tests in the first place.

I think we should definitely tag the cases which are errors (I think at least the remaining waste.1 after `opam install` is an error??) and we probably can't have too many `ls` calls in this test!

File: tests/reftests/clean.test (hunk 3 above)
Comment (rjbou) [reply tests/reftests/clean.test:191]: yes

**Review by kit-ty-kate — APPROVED** [review]

**Comment by kit-ty-kate** [comment]
Thanks!

**Pull request merged**